File: core/src/net/brenig/pixelescape/game/entity/impl/particle/EntityFadingParticle.java
package net.brenig.pixelescape.game.entity.impl.particle;

import com.badlogic.gdx.graphics.Color;

import net.brenig.pixelescape.PixelEscape;
import net.brenig.pixelescape.game.data.GameMode;
import net.brenig.pixelescape.game.entity.Entity;
import net.brenig.pixelescape.lib.Utils;
import net.brenig.pixelescape.render.WorldRenderer;

public class EntityFadingParticle extends Entity {

	private static final int size = 4;
	private static final int radius = size / 2;

	private float xVel = 0;
	private float yVel = 0;

	private float fadeDuration = 0.5F;
	private float fadeTimePassed = 0;

	private float xAccelerationFactor = 1F;
	private float yAccelerationFactor = 1F;

	private float color_r = 0;
	private float color_g = 0;
	private float color_b = 0;

	public void setColor(float color_r, float color_g, float color_b) {
		this.color_r = color_r;
		this.color_g = color_g;
		this.color_b = color_b;
	}

	public void setColor(Color color) {
		setColor(color.r, color.g, color.b);
	}

	public void setFadeDuration(float fadeDuration) {
		this.fadeDuration = fadeDuration;
	}

	public void setAccelerationFactor(float xAcc, float yAcc) {
		this.xAccelerationFactor = xAcc;
		this.yAccelerationFactor = yAcc;
	}

	public void setVelocity(float xVel, float yVel) {
		this.xVel = xVel;
		this.yVel = yVel;
	}

	@Override
	public void render(PixelEscape game, WorldRenderer renderer, GameMode gameMode, float delta) {
		if(isDead()) {
			return;
		}
		//Move
		xPos += xVel * delta;
		yPos += yVel * delta;

		xVel = Math.min(gameMode.getMaxEntitySpeed(), xVel * xAccelerationFactor);
		yVel = Math.min(gameMode.getMaxEntitySpeed(), yVel * yAccelerationFactor);

		fadeTimePassed += delta;

		float currentAlpha = 1 - Utils.easeInAndOut(fadeTimePassed, fadeDuration);

		game.getRenderManager().begin();
		game.getRenderManager().setColor(color_r, color_g, color_b, currentAlpha);
		renderer.renderRectWorld(xPos - radius, yPos - radius, size, size);
	}

	@Override
	public boolean isDead() {
		return fadeTimePassed >= fadeDuration;
	}


	@Override
	public void reset() {
		super.reset();
		color_r = 0;
		color_g = 0;
		color_b = 0;
		xVel = 0;
		yVel = 0;
		fadeDuration = 0.5F;
		fadeTimePassed = 0;
		xAccelerationFactor = 1F;
		yAccelerationFactor = 1F;
	}
}

File: core/src/net/brenig/pixelescape/game/entity/impl/particle/EntityFadingParticle.kt
package net.brenig.pixelescape.game.entity.impl.particle

import com.badlogic.gdx.graphics.Color
import net.brenig.pixelescape.PixelEscape
import net.brenig.pixelescape.game.data.GameMode
import net.brenig.pixelescape.game.entity.Entity
import net.brenig.pixelescape.lib.utils.AnimationUtils
import net.brenig.pixelescape.render.WorldRenderer

class EntityFadingParticle : Entity() {

    private var xVel = 0f
    private var yVel = 0f

    private var fadeDuration = 0.5f
    private var fadeTimePassed = 0f

    private var xAccelerationFactor = 1f
    private var yAccelerationFactor = 1f

    private var colorRed = 0f
    private var colorGreen = 0f
    private var colorBlue = 0f

    override val isDead: Boolean
        get() = fadeTimePassed >= fadeDuration

    fun setColor(colorRed: Float, colorGreen: Float, colorBlue: Float) {
        this.colorRed = colorRed
        this.colorGreen = colorGreen
        this.colorBlue = colorBlue
    }

    fun setColor(color: Color) {
        setColor(color.r, color.g, color.b)
    }

    fun setFadeDuration(fadeDuration: Float) {
        this.fadeDuration = fadeDuration
    }

    fun setAccelerationFactor(xAcc: Float, yAcc: Float) {
        this.xAccelerationFactor = xAcc
        this.yAccelerationFactor = yAcc
    }

    fun setVelocity(xVel: Float, yVel: Float) {
        this.xVel = xVel
        this.yVel = yVel
    }

    override fun render(game: PixelEscape, renderer: WorldRenderer, gameMode: GameMode, delta: Float) {
        if (isDead) {
            return
        }
        //Move
        xPos += xVel * delta
        yPos += yVel * delta

        xVel = Math.min(gameMode.maxEntitySpeed, xVel * xAccelerationFactor)
        yVel = Math.min(gameMode.maxEntitySpeed, yVel * yAccelerationFactor)

        fadeTimePassed += delta

        val currentAlpha = 1 - AnimationUtils.easeInAndOut(fadeTimePassed, fadeDuration)

        game.renderManager.begin()
        game.renderManager.setColor(colorRed, colorGreen, colorBlue, currentAlpha)
        renderer.renderRectWorld(xPos - radius, yPos - radius, size.toFloat(), size.toFloat())
    }


    override fun reset() {
        super.reset()
        colorRed = 0f
        colorGreen = 0f
        colorBlue = 0f
        xVel = 0f
        yVel = 0f
        fadeDuration = 0.5f
        fadeTimePassed = 0f
        xAccelerationFactor = 1f
        yAccelerationFactor = 1f
    }

    companion object {

        private const val size = 4
        private const val radius = size / 2
    }
}
